src/core/Akka.Persistence/Internal/CollectionsExtensions.cs

```diff
@@ -0,0 +1,22 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="CollectionsExtensions.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System.Collections.Immutable;
+
+namespace Akka.Persistence.Internal
+{
+    internal static class CollectionsExtensions
+    {
+        public static ImmutableArray<T> AddFirst<T>(this ImmutableArray<T> @this, T item)
```

Comment:
Have you checked the performance? AFAIK LinkedList AddFirst operation is O(1). Using ImmutableArray is copying the whole array into new struct: [implementation](https://github.com/dotnet/corefx/blob/29d7b6dee194e71bc5c98eec962bcac39bee8bb7/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs#L298).